src/transport/keepalive.rs

```diff
@@ -0,0 +1,199 @@
+use std::io;
+use std::time::Duration;
+use tokio::net::TcpStream;
+
+pub trait SetKeepalive {
+    fn keepalive(&self) -> io::Result<Option<Duration>>;
+    fn set_keepalive(&mut self, ka: Option<Duration>) -> ::std::io::Result<()>;
+}
+
+impl SetKeepalive for TcpStream {
+    fn keepalive(&self) -> io::Result<Option<Duration>> {
+        TcpStream::keepalive(self)
+    }
+
+    fn set_keepalive(&mut self, ka: Option<Duration>) -> ::std::io::Result<()> {
+        TcpStream::set_keepalive(self, ka)
+    }
+}
+
+pub mod accept {
+    use std::time::Duration;
+    use tokio::io::{AsyncRead, AsyncWrite};
+
+    use super::SetKeepalive;
+    use svc;
+
+    pub fn layer(keepalive: Option<Duration>) -> Layer {
+        Layer { keepalive }
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct Layer {
+        keepalive: Option<Duration>,
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct Stack<M> {
+        keepalive: Option<Duration>,
+        inner: M,
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct Accept<T> {
+        keepalive: Option<Duration>,
+        inner: T,
+    }
+
+    impl<T, M> svc::Layer<T, T, M> for Layer
+    where
+        M: svc::Stack<T>,
+    {
+        type Value = <Stack<M> as svc::Stack<T>>::Value;
+        type Error = <Stack<M> as svc::Stack<T>>::Error;
+        type Stack = Stack<M>;
+
+        fn bind(&self, inner: M) -> Self::Stack {
+            Stack {
+                inner,
+                keepalive: self.keepalive,
+            }
+        }
+    }
+
+    impl<T, M> svc::Stack<T> for Stack<M>
+    where
+        M: svc::Stack<T>,
+    {
+        type Value = Accept<M::Value>;
+        type Error = M::Error;
+
+        fn make(&self, target: &T) -> Result<Self::Value, Self::Error> {
```

Comment:
TIOLI: it seems like we could probably reduce some duplicate code (the `Layer` and `Stack` implementations) if this was implemented similarly to the `strip_header` layer, using a marker to indicate whether we're building an `Accept` or a `Connect:
https://github.com/linkerd/linkerd2-proxy/blob/d3e7f8bb69c0f2ed58bcbedc8acecf3079344e9b/src/proxy/http/strip_header.rs#L33-L83

Reply:
I find that approach a bit harder to work with, personally.  The overhead of having two similarly-shaped types is preferable, in my opinion, to having to reason about how a single type changes behavior in various situations. As it is, it's pretty clear and unambiguous, if slightly more verbose.